src/Tmds.Ssh/KerberosCredential.cs

```diff
@@ -0,0 +1,339 @@
+        if (!isAuthSuccess)
+        {
+            return false;
+        }
+
+        try
+        {
+            using var message = authContext.IsSigned
+                ? CreateGssapiMicData(connection.SequencePool, connectionInfo.SessionId!,  userName, authContext)
+                : CreateGssapiCompleteMessage(connection.SequencePool);
+            await connection.SendPacketAsync(message.Move(), ct).ConfigureAwait(false);
+        }
+        catch (MissingMethodException)
+        {
+            // Remove once .NET 8 is no longer the minimum and we get rid of reflection.
+            return false;
+        }
+
+        return true;
+    }
+
+    private static async Task<bool> TryStageOid(SshConnection connection, ILogger logger, string userName, CancellationToken ct)
+    {
+        {
+            using var userAuthMsg = CreateOidRequestMessage(connection.SequencePool,
+                                        userName, KRB5_OID);
+            await connection.SendPacketAsync(userAuthMsg.Move(), ct).ConfigureAwait(false);
+        }
+
+        while (true)
+        {
+            using Packet response = await connection.ReceivePacketAsync(ct).ConfigureAwait(false);
+            MessageId? mid = response.MessageId;
+
+            if (mid == MessageId.SSH_MSG_USERAUTH_BANNER)
+            {
+                // First message in exchange could be a banner which we ignore
+                // for now.
+                continue;
+            }
+
+            if (mid == MessageId.SSH_MSG_USERAUTH_FAILURE)
+            {
+                logger.AuthenticationKerberosFailed("No OID response received");
+                return false;
+            }
+            else if (mid != MessageId.SSH_MSG_USERAUTH_GSSAPI_RESPONSE)
+            {
+                ThrowHelper.ThrowProtocolUnexpectedValue();
+            }
+
+            ReadOnlySequence<byte> oidResponse = GetGssapiOidResponse(response);
+            if (KRB5_OID.AsSpan().SequenceEqual(oidResponse.IsSingleSegment ? oidResponse.FirstSpan : oidResponse.ToArray()))
+            {
+                return true;
+            }
+            else
+            {
+                var reader = new AsnReader(oidResponse.ToArray(), AsnEncodingRules.DER);
+                string receivedOid = reader.ReadObjectIdentifier();
+                logger.AuthenticationKerberosFailed($"OID response {receivedOid} did not match expected value");
+                return false;
+            }
+        }
+    }
+
+    private static async Task<bool> TryStageAuthentication(SshConnection connection, ILogger logger, NegotiateAuthentication authContext, CancellationToken ct)
+    {
+        byte[]? outToken;
+        NegotiateAuthenticationStatusCode statusCode;
+        try
+        {
+            outToken = authContext.GetOutgoingBlob(Array.Empty<byte>(), out statusCode);
+        }
+        catch (InvalidOperationException)
+        {
+            logger.AuthenticationKerberosFailed("Kerberos library not available");
```

Comment:
I think we might get `NegotiateAuthenticationStatusCode.Unsupported` when the Kerberos library is not available.

Did you try and see an `InvalidOperationException`?

Reply:
Looks like you are right. I was looking at https://github.com/dotnet/runtime/blob/6e88cf8c62a8100c36eef4bcc26a3d830b60ff35/src/libraries/System.Net.Security/src/System/Net/NegotiateAuthenticationPal.Unsupported.cs#L52-L56 and saw the `InvalidOperationException` but didn't realise `GetOutgoingBlock` was doing something different.

I'll update the code to handle this.